import.go

```diff
@@ -44,17 +59,107 @@ func newImporter(tree *MutableTree, version int64) (*Importer, error) {
 		return nil, errors.New("tree must be empty")
 	}
 
-	return &Importer{
-		tree:    tree,
-		version: version,
-		batch:   tree.ndb.db.NewBatch(),
-		stack:   make([]*Node, 0, 8),
-	}, nil
+	importer := &Importer{
+		tree:         tree,
+		version:      version,
+		batch:        tree.ndb.db.NewBatch(),
+		stack:        make([]*Node, 0, 8),
+		batchMtx:     sync.RWMutex{},
+		chBatch:      make(chan db.Batch, 1),
+		chBatchWg:    sync.WaitGroup{},
+		chNode:       make(chan *Node, desiredBatchSize),
+		chNodeWg:     sync.WaitGroup{},
+		chNodeData:   make(chan NodeData, desiredBatchSize),
+		chNodeDataWg: sync.WaitGroup{},
+	}
+
+	importer.chNodeWg.Add(1)
+	go serializeNode(importer)
+
+	importer.chNodeDataWg.Add(1)
+	go setBatchData(importer)
+
+	importer.chBatchWg.Add(1)
+	go batchWrite(importer)
+
+	return importer, nil
+
+}
+
+// serializeNode get the next validated and hashed node from the channel
+// and dump the node to byte buf, push the serialized node data to another channel
+func serializeNode(i *Importer) {
+	for i.batch != nil {
+		if currNode, open := <-i.chNode; open {
+			buf := bufPool.Get().(*bytes.Buffer)
+			buf.Reset()
+			if err := currNode.writeBytes(buf); err != nil {
+				panic(err)
+			}
+			bytesCopy := make([]byte, buf.Len())
+			copy(bytesCopy, buf.Bytes())
+			bufPool.Put(buf)
+
+			i.chNodeData <- NodeData{
+				node: currNode,
+				data: bytesCopy,
+			}
+		} else {
+			break
+		}
+	}
+	i.chNodeWg.Done()
+}
+
+// setBatchData get the next serialized node data from channel, and write the data to the current batch
+func setBatchData(i *Importer) {
+	for i.batch != nil {
+		if nodeData, open := <-i.chNodeData; open {
+			i.batchMtx.RLock()
+			if i.batch != nil {
+				err := i.batch.Set(i.tree.ndb.nodeKey(nodeData.node.hash), nodeData.data)
+				if err != nil {
+					panic(err)
+				}
+			}
+			i.batchMtx.RUnlock()
+			i.batchSize++
+			// Only commit a new batch if size meet desiredBatchSize and there's no pending batch write
+			if i.batchSize >= desiredBatchSize && len(i.chBatch) < 1 {
```

Comment:
If I/O is blocked for any reason then the batch could inflate (in memory) to the size of the remaining dataset to import (on disk) right?

Reply:
That is a valid concern, we can add back max batch size to prevent unbounded memory usage, so batch size would be controlled between desired size and max size, what do you think?